rasa/engine/training/graph_trainer.py

```diff
@@ -0,0 +1,222 @@
+
+        Walks the graph starting at each target node. If a node has a cache hit we
+        replace it with a `CachedComponent` and remove its input dependencies.
+        At the end, any node that is not an ancestor of a target node will be pruned
+        when we call `minimal_graph_schema()`.
+
+        Args:
+            schema: The graph to prune.
+            fingerprint_run_outputs: Node outputs from the fingerprint run as a mapping
+                from node name to output.
+
+        Returns:
+            The pruned schema.
+        """
+        pruned_schema = copy.deepcopy(schema)
+        target_node_names = pruned_schema.target_names
+
+        for target_node_name in target_node_names:
+            self._walk_and_prune(
+                pruned_schema, target_node_name, fingerprint_run_outputs
+            )
+
+        return pruned_schema.minimal_graph_schema()
+
+    def _walk_and_prune(
+        self,
+        schema: GraphSchema,
+        current_node_name: Text,
+        fingerprint_run_outputs: Dict[Text, Union[FingerprintStatus, Any]],
+    ) -> None:
+        """Recursively walks backwards though a graph checking the status of each node.
+
+        If node has a fingerprint key hit then we check if there is a cached output.
+        If there is a cached output we will replace the node with a `CachedComponent`
+        and remove all its dependencies (`.needs`). If there is not a fingerprint key
+        hit, or there is no cached output, the node is left untouched and will executed
+        again next run unless it is no longer the ancestor of a target node.
+
+        Args:
+            schema: The graph we are currently walking.
+            current_node_name: The current node on the walk.
+            fingerprint_run_outputs: The fingerprint statuses of every node as a mapping
+                from node name to status.
+        """
+        fingerprint_run_output = fingerprint_run_outputs[current_node_name]
+        node = schema.nodes[current_node_name]
+
+        # If we have replaced this node with a `CachedComponent` we have already
+        # visited this node. A CachedComponent is updated to have no parent nodes, so
+        # we can end the walk here.
+        if node.uses == CachedComponent:
+            return
+
+        # If the output was a `FingerprintStatus` we must check the cache and status.
+        if isinstance(fingerprint_run_output, FingerprintStatus):
+            # If there is a fingerprint key hit we can potentially use a cached output.
+            if fingerprint_run_output.is_hit:
+                output_result = self._cache.get_cached_result(
+                    output_fingerprint_key=fingerprint_run_output.output_fingerprint,
+                    node_name=current_node_name,
+                    model_storage=self._model_storage,
+                )
+                if output_result:
+                    logger.debug(f"Updating {node} to use a `CachedComponent`.")
```

Comment:
do we need to use `node_name` instead of `node` for a nicer print or will the dataclass look readable enough?

Reply:
schema looks ok, but will change to node name.